### PR TITLE
feat: add facebook provider

### PR DESCRIPTION
**Pull request by WindSpiritSR**
Fix: https://github.com/casbin/casdoor/issues/64

Deployed at: https://casdoor.leviatan.cn

Signed-off-by: WindSpirit <simon343riley@gmail.com>

### CODE REVIEW

File: object/user.go [hunk 1]
```diff
@@ -15,198 +15,199 @@
...
+    IsGlobalAdmin bool     `json:"isGlobalAdmin"`
+    IsForbidden   bool     `json:"isForbidden"`
+    Hash          string   `xorm:"varchar(100)" json:"hash"`
+    PreHash       string   `xorm:"varchar(100)" json:"preHash"`
+
+    Github string `xorm:"varchar(100)" json:"github"`
+    Google string `xorm:"varchar(100)" json:"google"`
+    QQ     string `xorm:"qq varchar(100)" json:"qq"`
+    WeChat string `xorm:"wechat varchar(100)" json:"wechat"`
+    Facebook string `xorm:"facebook varchar(100)" json:"facebook"`
```
Comment (hsluoyz) [inline object/user.go:52]: align

File: object/user.go [hunk 2]
```diff
@@ -15,198 +15,199 @@
...
+    Properties map[string]string `json:"properties"`
 }
 
 func GetGlobalUsers() []*User {
-	users := []*User{}
-	err := adapter.Engine.Desc("created_time").Find(&users)
-	if err != nil {
-		panic(err)
-	}
+    users := []*User{}
```
Comment (hsluoyz) [inline object/user.go:58]: We use 1 tab as indentation.

File: idp/facebook.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    token := oauth2.Token{
+        AccessToken: facebookAccessToken.AccessToken,
+        TokenType:   "FacebookAccessToken",
+        Expiry:      time.Time{},
+    }
+
+    return &token, nil
+}
+
+//{
```
Comment (hsluoyz) [inline idp/facebook.go:109]: Is this real response?

File: idp/facebook.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        return nil, err
+    }
+
+    if err = json.Unmarshal([]byte(userInfoResp), &facebookUserInfo); err != nil {
+        return nil, err
+    }
+
+    userInfo := UserInfo{
+        Id:          facebookUserInfo.Id,
+        Username:    facebookUserInfo.Name,
```
Comment (hsluoyz) [inline idp/facebook.go:167]: should be Id

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: idp/facebook.go (hunk 3 above)
Comment (WindSpiritSR) [reply idp/facebook.go:109]: The format of the real response looks like this, and the value of `id` `name` `url` and `email` value is modified here

**Review by WindSpiritSR — COMMENTED** [review]
fixed

**Comment by hsluoyz** [comment]
@WindSpiritSR any demo to try?

image

**Review by hsluoyz — APPROVED** [review]

**Comment by WindSpiritSR** [comment]
Application: https://casdoor.leviatan.cn/login/oauth/authorize?client_id=00c5f3410de2ebf737e4&response_type=code&redirect_uri=https://casdoor.leviatan.cn/callback&scope=read&state=casdoor
Facebook test email: `nngncxrqjb_1622875301@tfbnw.net`
Facebook test passwd: `fbtest123`

It will redirect to `https://www.leviatan.cn` if successful login with facebook account

**Comment by hsluoyz** [comment]
@WindSpiritSR , thanks, merged

**Pull request merged**